servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,648 @@
+   */
+  private final IndexedSet<MasterWorkerInfo> mWorkers =
+      new IndexedSet<MasterWorkerInfo>(mIdIndex, mAddressIndex);
+  /**
+   * Keeps track of workers which are no longer in communication with the master. Access must be
+   * synchronized on mWorkers.
+   */
+  private final BlockingQueue<MasterWorkerInfo> mLostWorkers =
+      new LinkedBlockingQueue<MasterWorkerInfo>();
+  /** The service that detects lost worker nodes, and tries to restart the failed workers. */
+  private Future<?> mLostWorkerDetectionService;
+  /** The next worker id to use. This state must be journaled. */
+  private final AtomicLong mNextWorkerId = new AtomicLong(1);
+
+  /**
+   * @param baseDirectory the base journal directory
+   * @return the journal directory for this master
+   */
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.BLOCK_MASTER_SERVICE_NAME);
+  }
+
+  public BlockMaster(Journal journal, TachyonConf tachyonConf) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("block-master-%d", true)));
+    mTachyonConf = tachyonConf;
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new BlockMasterService.Processor<BlockMasterServiceHandler>(
+        new BlockMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getServiceName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    // clear state before processing checkpoint.
+    mBlocks.clear();
+    super.processJournalCheckpoint(inputStream);
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    if (entry instanceof BlockIdGeneratorEntry) {
+      mBlockIdGenerator.setNextContainerId(((BlockIdGeneratorEntry) entry).getNextContainerId());
+    } else if (entry instanceof WorkerIdGeneratorEntry) {
+      mNextWorkerId.set(((WorkerIdGeneratorEntry) entry).getNextWorkerId());
+    } else if (entry instanceof BlockInfoEntry) {
+      BlockInfoEntry blockInfoEntry = (BlockInfoEntry) entry;
+      mBlocks.put(blockInfoEntry.getBlockId(),
+          new MasterBlockInfo(blockInfoEntry.getBlockId(), blockInfoEntry.getLength()));
+    } else {
+      throw new IOException("unexpected entry in journal: " + entry);
+    }
+  }
+
+  @Override
+  public void streamToJournalCheckpoint(JournalOutputStream outputStream) throws IOException {
+    outputStream.writeEntry(mBlockIdGenerator.toJournalEntry());
+    outputStream.writeEntry(new WorkerIdGeneratorEntry(mNextWorkerId.get()));
+    for (MasterBlockInfo blockInfo : mBlocks.values()) {
+      outputStream.writeEntry(new BlockInfoEntry(blockInfo.getBlockId(), blockInfo.getLength()));
+    }
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    super.start(isLeader);
+    if (isLeaderMode()) {
```

Comment:
it is strange that line 183 uses `isLeader` but line 184 uses `isLeaderMode()` to decide whether this master is leader or not. Is it possible that  they become inconsistent? Better to use a single source of truth